From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>
Date: Tue, 7 Jan 2025 13:36:09 +0100
Subject: [PATCH 14/29] Run the GC reconciliation multiple times if needed

The GC implementation will be split into 2 phases, so the Reconcile() function
will have to be called multiple times in the unit test to reach the final state.

The RequeueAfter value will be used to determine if the reconciliation should be
called again. The tests should call the Reconcile() function again as
long as the RequeueAfter value is less than the SyncPeriod.

For the current Garbage Collector implementation, the Reconcile() function
will be called once just like before.

Later, after the fix, the Reconcile() function will be called twice:
- the first call will mark the artifacts for deletion
- the second call will delete the artifacts.

Before the second call, the fake clock has to be advanced to allow to properly
check the time based conditions (CreationTimestamp, UnusedAt) in the Garbage
Collector.

The `labeledSecret9` and `labeledConfigMap5` objects test the aspect of the
Garbage Collector that recently created objects are not deleted. If the
Reconcile function is called multiple times and the fake clock is advanced, the
`labeledSecret9` and `labeledConfigMap5` objects need to be faked to be recently
created again and again, by advancing their CreationTimestamps in the fake
client.
---
 .../garbagecollector/reconciler_test.go       | 46 ++++++++++++++++---
 1 file changed, 40 insertions(+), 6 deletions(-)

diff --git a/pkg/resourcemanager/controller/garbagecollector/reconciler_test.go b/pkg/resourcemanager/controller/garbagecollector/reconciler_test.go
index 81e85aa34f..f14d5dbafc 100644
--- a/pkg/resourcemanager/controller/garbagecollector/reconciler_test.go
+++ b/pkg/resourcemanager/controller/garbagecollector/reconciler_test.go
@@ -162,8 +162,18 @@ var _ = Describe("Collector", func() {
 			Expect(c.List(ctx, configMapList)).To(Succeed())
 			Expect(configMapList.Items).To(BeEmpty())
 
-			_, err := gc.Reconcile(ctx, reconcile.Request{})
-			Expect(err).NotTo(HaveOccurred())
+			runs := 0
+			for {
+				result, err := gc.Reconcile(ctx, reconcile.Request{})
+				runs++
+				Expect(err).NotTo(HaveOccurred())
+				if result.Requeue && result.RequeueAfter < gc.Config.SyncPeriod.Duration {
+					fakeClock.Step(result.RequeueAfter)
+				} else {
+					break
+				}
+			}
+			Expect(runs).To(Equal(1))
 
 			secretList = &corev1.SecretList{}
 			Expect(c.List(ctx, secretList)).To(Succeed())
@@ -186,8 +196,18 @@ var _ = Describe("Collector", func() {
 			Expect(c.List(ctx, configMapList)).To(Succeed())
 			Expect(configMapList.Items).To(ConsistOf(*unlabeledConfigMap))
 
-			_, err := gc.Reconcile(ctx, reconcile.Request{})
-			Expect(err).NotTo(HaveOccurred())
+			runs := 0
+			for {
+				result, err := gc.Reconcile(ctx, reconcile.Request{})
+				runs++
+				Expect(err).NotTo(HaveOccurred())
+				if result.Requeue && result.RequeueAfter < gc.Config.SyncPeriod.Duration {
+					fakeClock.Step(result.RequeueAfter)
+				} else {
+					break
+				}
+			}
+			Expect(runs).To(Equal(1))
 
 			secretList = &corev1.SecretList{}
 			Expect(c.List(ctx, secretList)).To(Succeed())
@@ -242,8 +262,22 @@ var _ = Describe("Collector", func() {
 			Expect(c.Create(ctx, &corev1.Pod{ObjectMeta: objectMetaFor("pod1", labeledSecret6, labeledConfigMap6)})).To(Succeed())
 			Expect(c.Create(ctx, &batchv1.CronJob{ObjectMeta: objectMetaFor("cronjob2", labeledSecret7, labeledConfigMap7)})).To(Succeed())
 
-			_, err := gc.Reconcile(ctx, reconcile.Request{})
-			Expect(err).NotTo(HaveOccurred())
+			runs := 0
+			for {
+				result, err := gc.Reconcile(ctx, reconcile.Request{})
+				runs++
+				Expect(err).NotTo(HaveOccurred())
+				if result.Requeue && result.RequeueAfter < gc.Config.SyncPeriod.Duration {
+					fakeClock.Step(result.RequeueAfter)
+					labeledSecret9.CreationTimestamp.Time = labeledSecret9.CreationTimestamp.Time.Add(result.RequeueAfter)
+					c.Update(ctx, labeledSecret9)
+					labeledConfigMap5.CreationTimestamp.Time = labeledConfigMap5.CreationTimestamp.Time.Add(result.RequeueAfter)
+					c.Update(ctx, labeledConfigMap5)
+				} else {
+					break
+				}
+			}
+			Expect(runs).To(Equal(1))
 
 			secretList = &corev1.SecretList{}
 			Expect(c.List(ctx, secretList)).To(Succeed())
